react.d.ts: Change union type to intersection type

Strate opened the pull request:
`type ReactInstance = Component<any, any> | Element` is wrong. Trying to accessing any `Element`'s property gives me a compilation error:
```
this.refs['refname'].scrollHeight // error TS2339: Property 'scrollHeight' does not exist on type 'Component<any, any> | Element'.
```
This happening because of [typescript union type spec](https://github.com/Microsoft/TypeScript/issues/805):

> The type A|B has a property P of type X|Y if A has a property P of type X and B has a property P of type Y.
*(from Properties section)*

Instead of using union type, there should be [intersection type](https://github.com/Microsoft/TypeScript/pull/3622), which completely fits to this case:

> The type A & B has a property P if A has a property P or B has a property P.
*(from Properties section)*

So, there should be `type ReactInstance =  Component<any, any> & Element`

See also #6205

johnnyreilly commented:
Please could you fix the tests?

@jbrantly @vsiao @pspeter3 could you take a look when done?  Cheers.

Strate commented:
I fixed the tests, and I reverted change of `ReactInstance` type in favor of change to type of `Component.refs` element. I think I should write test for that, but I can not find out how to do that. Could anyone to help me with test?

adidahiya commented on react/react.d.ts:
```diff
@@ -125,7 +125,7 @@ declare namespace __React {
         context: {};
         refs: {
-            [key: string]: ReactInstance
+            [key: string]: Component<any, any> & Element
```
this is wrong; it should not be an intersection type; union was the right type. you have to narrow the type from `Component<any, any> | Element` to just `Element` to get your code to work

pspeter3 replied:
I agree. This is definitely a type union

adidahiya commented:
:-1:

johnnyreilly commented:
Closed based on feedback

Strate commented:
ahh, I got it. It really can be either component or element, for base react elements, such as `div` element. Sorry for disturb, guys.

johnnyreilly commented:
No problem
